### ISSUE TITLE
Swift as external storage downloads all files every time in NC11

### ISSUE THREAD

**Issue by jmdeboer-surfsara** [nextcloud/server#5868]
### Steps to reproduce
1. Install Nextcloud 11, use Swift as external storage
2. Upload some files
3. Connect using the desktop client

### Expected behaviour
Only new files should be downloaded

### Actual behaviour
All files are  downloaded every time.

### Server configuration

**Operating system**:  CentOS 7
**Web server:** Apache
**Database:** MySQL
**PHP version:** 7.0.20
**Nextcloud version:** 11.0.3
**Updated from an older Nextcloud/ownCloud or fresh install:** Fresh install
**Where did you install Nextcloud from:** tarball
**Signing status:** No errors have been found.

**List of activated apps:**
Enabled:
  - activity: 2.4.1
  - comments: 1.1.0
  - dav: 1.1.1
  - federatedfilesharing: 1.1.1
  - federation: 1.1.1
  - files: 1.6.1
  - files_external: 1.1.2
  - files_pdfviewer: 1.0.1
  - files_sharing: 1.1.1
  - files_texteditor: 2.2
  - files_trashbin: 1.1.0
  - files_versions: 1.4.0
  - files_videoplayer: 1.0.0
  - gallery: 16.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.0
  - notifications: 1.0.1
  - password_policy: 1.1.0
  - provisioning_api: 1.1.0
  - sharebymail: 1.0.1
  - systemtags: 1.1.3
  - theming: 1.1.1
  - twofactor_backupcodes: 1.0.0
  - updatenotification: 1.1.1
  - workflowengine: 1.1.1
Disabled:
  - admin_audit
  - encryption
  - external
  - files_accesscontrol
  - files_automatedtagging
  - files_retention
  - firstrunwizard
  - logreader
  - serverinfo
  - survey_client
  - templateeditor
  - user_external
  - user_ldap
  - user_saml

**Nextcloud configuration:**
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***",
        ],
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "overwritehost": "***REMOVED SENSITIVE VALUE***l",
        "overwriteprotocol": "https",
        "forwarded_for_headers": [
            "HTTP_X_FORWARDED_FOR",
            "HTTP_FORWARDED_FOR"
        ],
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379,
            "timeout": 0,
            "password": "***REMOVED SENSITIVE VALUE***",
            "dbindex": 0
        },
        "filelocking.enabled": "true",
        "trusted_proxies": [
            "145.100.5.230"
        ],
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "version": "11.0.3.2",
        "appcodechecker": false,
        "updatechecker": false,
        "installed": true,
        "enable_previews": true,
        "preview_max_x": 1024,
        "preview_max_y": 1024,
        "preview_max_filesize_image": 100,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown"
        ],
        "appstoreenabled": false,
        "has_internet_connection": false,
        "auth.bruteforce.protection.enabled": true,
        "enable_avatars": true,
        "dbtype": "mysql",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "cron_log": true,
        "log_type": "file",
        "logfile": "\/var\/log\/***REMOVED SENSITIVE VALUE***.log",
        "logtimezone": "Europe\/Amsterdam",
        "loglevel": "3",
        "logdateformat": "d\/M\/Y:H:i:s",
        "objectstore_multibucket": {
            "class": "\\OC\\Files\\ObjectStore\\Swift",
            "arguments": {
                "autocreate": true,
                "username": "***REMOVED SENSITIVE VALUE***",
                "password": "***REMOVED SENSITIVE VALUE***",
                "region": "RegionOne",
                "url": "***REMOVED SENSITIVE VALUE***",
                "tenantName": "***REMOVED SENSITIVE VALUE***",
                "serviceName": "Swift",
                "urlType": "internal"
            }
        }
    }
}

**Are you using external storage, if yes which one:** Swift
**Are you using encryption:** no
**Are you using an external user-backend, if yes which one:** None

### Client configuration
**Browser:** Desktop client 2.3.2

**Operating system:** Ubuntu linux, Windows

**Comment by jmdeboer-surfsara** [comment]
The problem in caused in this file:
apps/files_external/lib/Lib/Storage/Swift.php
in the function hasUpdated().

This compares a cached directory listing with an actual directory listing.

When looking at the contents of the two arrays, I noticed that one ($content) has an entry for the current directory '.', whereas the other ($cachedNames) does not. So these arrays are never the same, and hasUpdated() always returns true.

A fix is this, at line 636, replace
while (($file = readdir($dh)) !== false) {
    $content[] = $file;
}

with 

while (($file = readdir($dh)) !== false) {
    if ($file !== '.') {
        $content[] = $file;
    }
}

After that, only new files are downloaded.

**Comment by MorrisJobke** [comment]
cc @rullzer @icewind1991 

@jmdeboer-surfsara Could you try Nextcloud 12, because we improved a lot there regarding the object storage as primary storage.

**Comment by MorrisJobke** [comment]
> @jmdeboer-surfsara Could you try Nextcloud 12, because we improved a lot there regarding the object storage as primary storage.

Ah, I just re-read it and it is mounted as external storage. Is this right? Because the config shows the entry for `objectstore_multibucket` and this would mean that it is used as primary storage.

**Comment by jmdeboer-surfsara** [comment]
Hi,

We use Swift both as primary and as external storage. The problem only seems to occur with the external storage. I am trying to figure out what happens exactly, and I noticed that when I create a new external storage, an object with a name of . (so dot) is placed in the container.

I created a new external storage, with the container name 'external2' and if I look at its contents:
```
$ swift list external2
.

$ swift stat external2
                  Account: ***hidden***
                Container: external2
                  Objects: 1
                    Bytes: 0
                 Read ACL:
                Write ACL:
                  Sync To:
                 Sync Key:
            Accept-Ranges: bytes
Strict-Transport-Security: max-age=15768000
         X-Storage-Policy: Standard-Replica
            Last-Modified: Tue, 25 Jul 2017 09:53:21 GMT
              X-Timestamp: 1500976400.07613
               X-Trans-Id: txec731ff81eaa4561889e9-00597716b0
             Content-Type: text/plain; charset=utf-8
   X-Openstack-Request-Id: txec731ff81eaa4561889e9-00597716b0

$ swift stat external2 "."
                  Account: ***hidden***
                Container: external2
                   Object: .
             Content Type: httpd/unix-directory
           Content Length: 0
            Last Modified: Tue, 25 Jul 2017 09:53:21 GMT
                     ETag: d41d8cd98f00b204e9800998ecf8427e
            Accept-Ranges: bytes
Strict-Transport-Security: max-age=15768000
              X-Timestamp: 1500976400.13686
               X-Trans-Id: txe90ddec0030247a1a9fb7-00597716d7
   X-Openstack-Request-Id: txe90ddec0030247a1a9fb7-00597716d7
```
So there is probably something going on with that object that I am not 100% aware of, it must be there for a reason ;) And therefore it is not impossible that my addition to hasUpdated() actually breaks something else.

**Comment by MorrisJobke** [comment]
> I noticed that when I create a new external storage, an object with a name of . (so dot) is placed in the container.

This one will be fixed in 11.0.5, 12.0.3 and 13.

**Comment by wiswedel** [comment]
Can this be closed? Please test again on 16.0.1

**Comment by MorrisJobke** [comment]
Should work in the meantime.
